File: src/targets/tls/chain_validity.py
"""Checks the TLS certificate chain validity for the Wire domain.

Uses openssl s_client -showcerts to verify the full certificate chain.
Incomplete chains work in some browsers but break on mobile or strict
TLS implementations.
"""

from __future__ import annotations

# Ours
from src.lib.base_target import BaseTarget
from src.lib.shell_safety import validate_domain_for_shell


class ChainValidity(BaseTarget):
    """Checks TLS certificate chain validity for the Wire domain.

    Connects via openssl from the admin host and inspects the verification
    result to determine if the chain is complete.
    """

    # TLS chain validity checks connect to public domain:443 requires external
    # internet access, not available from the internal admin host.
    requires_external_access: bool = True

    # Uses SSH to admin host for openssl checks
    requires_ssh: bool = True

    @property
    def description(self) -> str:
        """Human-readable description of what this target checks."""
        return "TLS certificate chain validity"

    @property
    def explanation(self) -> str:
        """Why this check exists and what determines healthy vs unhealthy."""
        return (
            "Incomplete certificate chains work in some browsers but break on mobile or "
            "strict TLS implementations. Good when openssl reports verify OK."
        )


    def collect(self) -> bool:
        """Check the TLS certificate chain via openssl s_client.

        Returns:
            True if the certificate chain is valid, False otherwise.

        Raises:
            RuntimeError: If the connection to the domain fails entirely.
        """
        domain: str = self.config.cluster.domain
        validate_domain_for_shell(domain)
        self.terminal.step(f"Checking TLS certificate chain for {domain}...")

        # openssl s_client outputs «Verify return code: 0 (ok)» on success
        cmd: str = (
            f"echo Q | openssl s_client -showcerts -connect {domain}:443"
            f" -servername {domain} 2>&1"
        )

        result = self.run_ssh(self.config.admin_host.ip, cmd)
        output: str = result.stdout

        if not output.strip():
            raise RuntimeError(f"Could not connect to {domain}:443 for TLS verification")

        # Count certificates in chain by counting BEGIN CERTIFICATE markers
        cert_count: int = output.count("BEGIN CERTIFICATE")

        # Check the verify return code 0 means the chain verified OK
        valid: bool = "Verify return code: 0 (ok)" in output

        # Extract the verify return code line for reporting
        verify_line: str = ""
        for line in output.split("\n"):
            if "Verify return code:" in line:
                verify_line = line.strip()
                break

        if valid:
            self._health_info = f"Chain valid, {cert_count} cert(s) in chain"
        else:
            self._health_info = f"Chain INVALID: {verify_line}, {cert_count} cert(s)"

        return valid

File: ui/src/checkers/tls/chain_validity.ts
/**
 * Makes sure the TLS certificate chain is complete and valid.
 *
 * The tls/chain_validity target tells us if the server is presenting
 * the full chain (root -> intermediate -> leaf). If it's incomplete,
 * desktop browsers might still work because they cache intermediates,
 * but mobile clients and strict validators will fail.
 */

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import type { DataLookup } from '../data_lookup'

export class ChainValidityChecker extends BaseChecker {
    readonly path: string = 'tls/chain_validity'
    readonly name: string = 'TLS certificate chain validity'
    readonly category: string = 'TLS / Certificates'
    readonly interest = 'Health' as const
    readonly explanation: string = 'Verifies the server presents a **complete TLS certificate chain** (root, intermediate, leaf). Incomplete chains may work in desktop browsers that cache intermediates but break mobile clients and strict TLS validators.'

    check(data: DataLookup): CheckResult {
        const point = data.get_applicable('tls/chain_validity') ?? data.get_applicable('direct/tls/chain_validity')

        // No data from either the SSH or direct (kubernetes) path
        if (!point) {
            // get_applicable() filters out not_applicable sentinels, so check
            // whether the SSH target was intentionally skipped (e.g. admin host)
            if (data.is_not_applicable('tls/chain_validity')) {
                return {
                    status: 'not_applicable',
                    status_reason: 'This check requires an external network perspective and was not applicable for this run.',
                    recommendation: 'This check requires running the gatherer from an internet-connected machine. Re-run with --source external to verify the TLS certificate chain from outside.',
                }
            }

            return {
                status: 'gather_failure',
                status_reason: 'TLS certificate chain validity data was not collected.',
                fix_hint: '1. Verify the gatherer script ran with network access to the target domain on port **443**.\n2. Re-run the gatherer and check its logs for connection errors:\n   ```\n   python3 src/script/runner.py --target tls/chain_validity\n   ```\n3. Ensure `openssl` is installed on the admin host.',
                recommendation: 'TLS certificate chain validity data not collected.',
            }
        }

        // Collection ran but the command failed
        if (point.value === null) {
            return {
                status: 'gather_failure',
                status_reason: 'TLS chain validity data was collected but contained no value.',
                recommendation: point.metadata?.error ?? 'TLS chain validity target ran but returned no result.',
                raw_output: point.raw_output,
            }
        }

        const val: string | number | boolean = point.value

        // If we got a string, check if it says the chain is broken
        if (typeof val === 'string') {
            const lower: string = val.toLowerCase()

            if (lower.includes('incomplete') || lower.includes('error')) {
                return {
                    status: 'unhealthy',
                    status_reason: 'Certificate chain reported as: "{{val}}".',
                    fix_hint: '1. Inspect the certificate chain served by the domain:\n   ```\n   openssl s_client -connect <domain>:443 -servername <domain> -showcerts </dev/null 2>&1\n   ```\n2. Verify the chain includes **root**, **intermediate**, and **leaf** certificates in the correct order.\n3. If intermediates are missing, concatenate them into the certificate file:\n   ```\n   cat leaf.crt intermediate.crt > fullchain.crt\n   ```\n4. If using cert-manager, check the `Certificate` resource for chain configuration issues:\n   ```\n   kubectl describe certificate <name> -n <namespace>\n   ```\n5. After fixing, verify the chain with: `openssl verify -CAfile ca-bundle.crt fullchain.crt`',
                    recommendation: 'TLS certificate chain is incomplete or invalid. Missing intermediates work in some browsers but break on mobile or strict clients.',
                    display_value: val,
                    raw_output: point.raw_output,
                    template_data: { val },
                }
            }

            // String with no error keywords, so chain is fine
            return {
                status: 'healthy',
                status_reason: 'Certificate chain validation returned: "{{val}}".',
                display_value: val,
                raw_output: point.raw_output,
                template_data: { val },
            }
        }

        // True means the chain is there and working
        if (val === true) {
            return {
                status: 'healthy',
                status_reason: 'Certificate chain is **complete and valid**.',
                display_value: val,
                raw_output: point.raw_output,
            }
        }

        // False or anything falsy means the chain is missing pieces
        return {
            status: 'unhealthy',
            status_reason: 'Certificate chain validation returned **false**, indicating an incomplete or invalid chain.',
            fix_hint: '1. Inspect the certificate chain served by the domain:\n   ```\n   openssl s_client -connect <domain>:443 -servername <domain> -showcerts </dev/null 2>&1\n   ```\n2. Verify the chain includes **root**, **intermediate**, and **leaf** certificates.\n3. If intermediates are missing, concatenate them into the certificate file:\n   ```\n   cat leaf.crt intermediate.crt > fullchain.crt\n   ```\n4. If using cert-manager, check the `Certificate` resource:\n   ```\n   kubectl describe certificate <name> -n <namespace>\n   ```\n5. After fixing, verify: `openssl verify -CAfile ca-bundle.crt fullchain.crt`',
            recommendation: 'TLS certificate chain is incomplete or invalid. Missing intermediates work in some browsers but break on mobile or strict clients.',
            display_value: val,
            raw_output: point.raw_output,
        }
    }
}

export default ChainValidityChecker
